File: 프로그래머스/구현/주차요금계산.py
import math

def getTotalTime(inTime, outTime):
    Intime = list(map(int, inTime.split(":")))
    Outtime = list(map(int, outTime.split(":")))
    return (Outtime[0]*60 + Outtime[1]) - (Intime[0]*60 + Intime[1])

def solution(fees, records):
    In = []
    totalTimes = {}
    for record in records:
        data = record.split(" ")
        if data[2] == "IN":
            In.append(data)
        else:
            inTime = ''
            tmp = []
            for i in In:
                if i[1] == data[1]:
                    inTime = i[0]
                else:
                    tmp.append(i)
            In = tmp
            term = getTotalTime(inTime, data[0])
            if data[1] in totalTimes:
                totalTimes[data[1]] += term
            else:
                totalTimes[data[1]] = term
    for data in In:
        term = getTotalTime(data[0], "23:59")
        if data[1] in totalTimes:
            totalTimes[data[1]] += term
        else:
            totalTimes[data[1]] = term
    ans = []
    for key,value in totalTimes.items():
        fee = fees[1] + math.ceil((value - fees[0])/fees[2]) * fees[3]
        if value <= fees[0]:
            fee = fees[1]
        ans.append([key, fee])
    return [x[1] for x in sorted(ans, key=lambda x:x[0])]

File: 자바스크립트/프로그래머스/주차요금계산.js
function convertToMin(time){
    const [hour, min] = time.split(":").map((x) => Number(x))
    return hour * 60 + min
}

function getParkFee(totalTime, fees){
    const [defaultTime, defaultFee, term, termFee] = fees;
    let tmp = totalTime - defaultTime;
    if (tmp <= 0) return defaultFee
    return defaultFee + Math.ceil(tmp / term) * termFee;
}

function solution(fees, records) {
    const parking = {}
    const totalParkingTime = {}
    for (const record of records){
        const [time, num, type] = record.split(" ")
        if (type === "IN"){
            parking[num] = time
        } else{
            const totalTime = convertToMin(time) - convertToMin(parking[num])
            delete parking[num]
            if (totalParkingTime[num]){
                totalParkingTime[num] += totalTime
            }else{
                totalParkingTime[num] = totalTime
            }
        }
    }
    for (const [key, value] of Object.entries(parking)){
        const totalTime = convertToMin('23:59') - convertToMin(value)
        if (totalParkingTime[key]){
            totalParkingTime[key] += totalTime
        }else{
            totalParkingTime[key] = totalTime
        }
    }
    return Object.entries(totalParkingTime)
        .sort((prev, next) => prev[0] - next[0])
        .map((item) => getParkFee(item[1], fees))
}